src/status_im2/common/keychain/events.cljs

```diff
@@ -86,26 +86,74 @@
   [key-uid password]
   (keychain/save-credentials key-uid key-uid (security/safe-unmask-data password) #()))
 
+(defn get-user-password!
+  [key-uid callback]
+  (keychain/get-credentials key-uid
+                            #(callback (when %
+                                         (-> %
+                                             (oops/oget "password")
+                                             (security/mask-data))))))
+
 (re-frame/reg-fx
  :keychain/get-user-password
  (fn [[key-uid callback]]
-   (keychain/get-credentials
-    key-uid
-    #(if % (callback (security/mask-data (oops/oget % "password"))) (callback nil)))))
+   (get-user-password! key-uid callback)))
 
 (rf/defn get-user-password
   [_ key-uid callback]
   {:keychain/get-user-password [key-uid callback]})
 
+(defn- password-migration-key-name
+  [key-uid]
+  (str key-uid "-password-migration"))
+
+(defn save-password-migration!
+  [key-uid]
+  (keychain/save-credentials
+   (password-migration-key-name key-uid)
+   key-uid
+   ;; NOTE: using the key-id as the password, but we don't really care about the
+   ;; value, we only care that it's there
+   key-uid
+   #(when-not %
+      (log/error
+       (str "Error while setting up keychain migration")))))
+
+(defn get-password-migration!
+  [key-uid callback]
+  (keychain/get-credentials
+   (password-migration-key-name key-uid)
+   (comp callback boolean)))
+
 (re-frame/reg-fx
  :keychain/clear-user-password
  (fn [key-uid]
+   (keychain/reset-credentials (password-migration-key-name key-uid))
    (keychain/reset-credentials key-uid)))
 
 (re-frame/reg-fx
  :keychain/save-password-and-auth-method
  (fn [{:keys [key-uid masked-password on-success on-error]}]
    (-> (save-user-password! key-uid masked-password)
        (.then #(save-auth-method! key-uid auth-method-biometric))
+       (.then #(save-password-migration! key-uid))
        (.then #(when on-success (on-success)))
        (.catch #(when on-error (on-error %))))))
+
+;; NOTE: migrating the plaintext password in the keychain
+;; with the hashed one. Added due to the sync onboarding
+;; flow, where the password arrives already hashed.
+(re-frame/reg-fx
+ :keychain/password-hash-migration
+ (fn [{:keys [key-uid callback]}]
+   (-> (get-password-migration! key-uid identity)
+       (.then (fn [migrated?]
+                (if migrated?
+                  (callback)
+                  (-> (get-user-password! key-uid identity)
+                      (.then #(security/hash-masked-password %))
+                      (.then #(save-user-password! key-uid %))
+                      (.then #(save-password-migration! key-uid))
+                      (.then #(callback))
```

Comment:
@cammellos @ilmotta I stumbled on the [promesa](https://funcool.github.io/promesa/latest/promises.html) package that has some really nice utilities for working with promises (both clj/cljs). It would be good for cases like this, where it can be simplified to smth like:
 ```cljs
(p/->> (get-user-password! key-uid identity)
        (security/hash-masked-password)
        (save-user-password! key-uid)
        ...)
```
Do you think it be a good addition to the project?

Reply:
@clauxx, I'm probably biased in favor of `promesa` because I extensively used it in the past. The library has a very solid API, so it's like one of those libs in Clojure where you add to the project and profit without maintenance burden.

OTOH, I used `promesa` in projects that heavily used promises everywhere (Node.js runtime instead of browser). Every code path had promises, so `promesa` paid off. In status-mobile we mostly have a few promises in re-frame effects (thankfully 😃). Therefore, our usages are very linear and simple, and I don't see `promesa` adding that much value.

I have also noticed promises code in status-mobile is quite convoluted and show some lack of knowledge of how promises work. Namespaces like `react-native.cameraroll` and `react-native.permissions`, among others show that to me because they confusingly mix callbacks with promises, probably an attempt to hide promises.

If we add `promesa` on top, I think it will be nice for some, but another layer on top of promises that some are not willing to learn.

In the end, I vote for adding it if you want in a separate PR, but better check with more devs.

Reply:
I saw the same issue in the `react-native.keychain` namespace, which is a bit annoying. Found myself pass `identity` as the callback in several places to make use of the resolved promise values. I'd say the value would be in having a consistent way of dealing with promises across the project and getting rid of the callbacks entirely, but  this could be done without the library by just adding it in the guideline.